designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+3. *Cluster Manager* that has a wide range of concerns including cluster create, upgrade and delete logic; network installation; AWS IAM management; OIDC management; GitOps management.
+
+Reverse engineering is time consuming and can lead to incorrect assumptions that result in incorrectly implemented behavior.
+
+### Tenets
+
+1. **Clarity**: clear and concise expectations produce a better developer experience.
+2. **Modular**: modular code that strives for single responsibility is easier to delete when requirements inevitably change.
+3. **Complimentary**: engineers have put time and effort into writing designs and implementing code so new code must take into consideration in-flight work and concrete future work.
+
+### Goals and objectives
+
+* Reduce the reliance by provider developers on reverse engineering workflows to understand expected behavior.
+* Disambiguate the installation and upgrade of provider components from cluster creation, upgrade and deletion.
+* Reduce the scope of workflows to provider component management (installation and upgrade) and cluster management (create, upgrade and delete).
+* Provide a mechanism to hook into and extend workflows with additional tasks during cluster creation, upgrade and deletion.
+* Reduce the surface area of the provider interface.
+
+### Statement of scope
+
+#### In Scope
+
+* Refactoring of workflow tasks to have a single responsibility.
+* Refactoring of cluster management components such that behavior can be invoked atomically in the context of a task. Behavior not capable of being broken into atomic actions complementary to tasks should be clearly documented against tasks.
+* Refactoring of parsing, defaulting and validating input specific to providers.
+* Refactoring of provider interfaces to align with workflow expectations (will require adjustment of provider implementations to satisfy the interfaces).
+* Documenting workflow tasks using native go docs so that provider developers have a clear understanding of the context in which they are executed.
+
+#### Out of Scope
+
+* Holistic refactoring of the cluster management component. The cluster management component has a hard to pin down scope of concerns that are heavily leveraged by workflows and it would take a broader effort to break them apart and organize that likely isn’t required for these proposed changes.
+* A plug-able provider interface that can be used to develop off-tree providers. Off-tree provider development requires much broader thought such as how our API should be structured to enable API extension, how to version the APIs and ensure they function together, and how to build and distribute artifacts specific to a given provider.
```

Comment:
can you clarify 'off-tree' in this context

Reply:
External to AWS repositories.

Will update.